google/cloud/log.h

```diff
@@ -0,0 +1,399 @@
+  virtual void ProcessWithOwnership(LogRecord log_record) = 0;
+};
+
+/**
+ *
+ */
+class LogSink {
+ public:
+  LogSink();
+
+  /// Return true if the severity is enabled at compile time.
+  constexpr static bool CompileTimeEnabled(Severity level) {
+    return level >= Severity::LOWEST_ENABLED;
+  }
+
+  /**
+   * Return the singleton instance for this application.
+   */
+  static LogSink& Instance();
+
+  /**
+   * Return true if this object has no backends.
+   *
+   * We want to avoid synchronization overhead when checking if a log message is
+   * enabled. Most of the time, most messages will be disabled, so incurring the
+   * locking overhead on each message would be too expensive and would
+   * discourage developers from creating logs. Furthermore, missing a few
+   * messages while the change of state "propagates" to other threads does not
+   * affect the correctness of the program.
+   *
+   * Note that `memory_order_relaxed` does not provide a compiler barrier
+   * either, so in theory stores into the atomic could be reordered by the
+   * optimizer. We have no reason to worry about that because all the writes
+   * are done inside a critical section protected by a mutex. The compiler
+   * cannot (or should not) reorder operations around those.
+   */
+  bool empty() const { return empty_.load(std::memory_order_relaxed); }
+
+  /**
+   * Return true if @p severity is enabled.
+   *
+   * We want to avoid synchronization overhead when checking if a log message is
+   * enabled. Most of the time, most messages will be disabled, so incurring the
+   * locking overhead on each message would be too expensive and would
+   * discourage developers from creating logs. Furthermore, missing a few
+   * messages while the change of state "propagates" to other threads does not
+   * affect the correctness of the program.
+   *
+   * Note that `memory_order_relaxed` does not provide a compiler barrier
+   * either, so in theory stores into the atomic could be reordered by the
+   * optimizer. We have no reason to worry about that because all the writes
+   * are done inside a critical section protected by a mutex. The compiler
+   * cannot (or should not) reorder operations around those.
+   */
+  bool is_enabled(Severity severity) const {
+    auto minimum = minimum_severity_.load(std::memory_order_relaxed);
+    return static_cast<int>(severity) >= minimum;
+  }
+
+  void set_minimum_severity(Severity minimum) {
+    minimum_severity_.store(static_cast<int>(minimum));
+  }
+  Severity minimum_severity() const {
+    return static_cast<Severity>(minimum_severity_.load());
+  }
+
+  long AddBackend(std::shared_ptr<LogBackend> backend);
+  void RemoveBackend(long id);
+  void ClearBackends();
+
+  void Log(LogRecord log_record);
+
+  /// Enable `std::clog` on `LogSink::Instance()`.
+  static long EnableStdClog();
+
+ private:
+  std::atomic<bool> empty_;
+  std::atomic<int> minimum_severity_;
+  std::mutex mu_;
```

Comment:
Can we use `GUARDED_BY` annotations on fields protected by `mu_` here, and have thread-safety analysis tell us if any code path is not properly locking the right mutex prior to accessing a protected member var?

https://clang.llvm.org/docs/ThreadSafetyAnalysis.html has more info.

Reply:
In principle, yes.  The easiest way would be to have the Abseil headers available, because otherwise we need to detect if `GUARDED_BY` is a thing.  Should I open a bug or make that part of this PR?

Reply:
+1.  I don't feel strongly about it being in this CL vs done in a future CL, but `GUARDED_BY` annotations would be great.